Thanks @drbenmorgan 

So it compiles and runs on the CPU, but compilation fails on the GPU for me:

```
titanx:~/build/AdePT$ ninja
[3/7] Building CXX object CMakeFiles/CopCore.dir/src/backend/Utils.cpp.o
FAILED: CMakeFiles/CopCore.dir/src/backend/Utils.cpp.o
/usr/bin/c++ -DTARGET_DEVICE_CUDA -I/home/graemes/code/AdePT/base/inc/CopCore/include -std=gnu++17 -MD -MT CMakeFiles/CopCore.dir/src/backend/Utils.cpp.o -MF CMakeFiles/CopCore.dir/src/backend/Utils.cpp.o.d -o CMakeFiles/CopCore.dir/src/backend/Utils.cpp.o -c /home/graemes/code/AdePT/base/inc/CopCore/src/backend/Utils.cpp
In file included from /home/graemes/code/AdePT/base/inc/CopCore/include/CopCore/backend/BackendCommon.h:25,
                 from /home/graemes/code/AdePT/base/inc/CopCore/src/backend/Utils.cpp:5:
/home/graemes/code/AdePT/base/inc/CopCore/include/CopCore/backend/CUDABackend.h:10:10: fatal error: cuda_runtime_api.h: No such file or directory
 #include <cuda_runtime_api.h>
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

I guess this was expected? We need to add a bit more plumbing for the CUDA setup.

--- a/base/inc/CopCore/CMakeLists.txt
+++ b/base/inc/CopCore/CMakeLists.txt
@@ -28,6 +28,7 @@ set_property(CACHE CUDA_ARCH PROPERTY STRINGS COMPATIBILITY MAX MIN 53 60 61 62 
 # Dependencies
 if(TARGET_DEVICE STREQUAL "CUDA")
   enable_language(CUDA)
+  find_package(CUDAToolkit 11.0 REQUIRED)
 endif()
 
 # Core library
@@ -52,6 +53,7 @@ target_compile_definitions(CopCore PUBLIC ${TARGET_DEFINITION})
 target_include_directories(CopCore PUBLIC
   $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
   $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
+target_link_libraries(CopCore PUBLIC $<$<STREQUAL:"${TARGET_DEVICE}","CUDA">:CUDA::cudart_static>)
 
 # Alias to assist in use as a subproject
 add_library(CopCore::CopCore ALIAS CopCore)

--- a/base/inc/CopCore/include/CopCore/Invoke.cuh
+++ b/base/inc/CopCore/include/CopCore/Invoke.cuh
@@ -20,20 +20,15 @@
  * @return     Return value of the function.
  */
 #if defined(TARGET_DEVICE_CPU) || (defined(TARGET_DEVICE_HIP) && (defined(__HCC__) || defined(__HIP__))) || \
-  ((defined(TARGET_DEVICE_CUDA) && defined(__CUDACC__)) || (defined(TARGET_DEVICE_CUDACLANG) && defined(__CUDA__)))
-template<class Fn, class Tuple, unsigned long... I>
-void invoke_impl(
-  Fn&& function,
-  const dim3& grid_dim,
-  const dim3& block_dim,
-  cudaStream_t stream,
-  const Tuple& invoke_arguments,
-  std::index_sequence<I...>)
+    ((defined(TARGET_DEVICE_CUDA) && defined(__CUDACC__)) || (defined(TARGET_DEVICE_CUDACLANG) && defined(__CUDA__)))
+template <class Fn, class Tuple, unsigned long... I>
+void invoke_impl(Fn &&function, const dim3 &grid_dim, const dim3 &block_dim, cudaStream_t stream,
+                 const Tuple &invoke_arguments, std::index_sequence<I...>)
 {
   // If any grid dimension component, or any block dimension component is zero,
   // return without running.
-  if (
-    grid_dim.x == 0 || grid_dim.y == 0 || grid_dim.z == 0 || block_dim.x == 0 || block_dim.y == 0 || block_dim.z == 0) {
+  if (grid_dim.x == 0 || grid_dim.y == 0 || grid_dim.z == 0 || block_dim.x == 0 || block_dim.y == 0 ||
+      block_dim.z == 0) {
     return;
   }
 
@@ -56,8 +51,8 @@ void invoke_impl(
 #endif
 }
 #else
-template<class Fn, class Tuple, unsigned long... I>
-void invoke_impl(Fn&&, const dim3&, const dim3&, cudaStream_t, const Tuple&, std::index_sequence<I...>)
+template <class Fn, class Tuple, unsigned long... I>
+void invoke_impl(Fn &&, const dim3 &, const dim3 &, cudaStream_t, const Tuple &, std::index_sequence<I...>)
 {
   error_cout << "Global function invoked with unexpected backend.\n";
 }

--- a/base/inc/CopCore/tests/test_saxpy.cu
+++ b/base/inc/CopCore/tests/test_saxpy.cu
@@ -42,7 +42,7 @@ int main()
 
   // This is the awkward part - how to set the grid/block appriopriately
   // for CPU vs GPU.
-  invoke_impl(saxpy, dim3((N+255)/256), dim3(256), cudaStream_t{}, args, indices{});
+  invoke_impl(saxpy, dim3((N + 255) / 256), dim3(256), cudaStream_t{}, args, indices{});
 
   // Check
 

